### DATA Wrangling

### <b> Problem Statement </b>
 
You are a junior data scientist and you are assigned a new task to perform data wrangling on a set of datasets. The datasets have many ambiguities. You have to identify those and apply different data wrangling techniques to get a dataset for further usage.

#### <b> Dataset </b>

- Download the `rental_bike_descr, rental_bike_season, and final_rental_bike_dataset` from Course Resources and upload the datasets to the lab

**Attribute Information:**


* date = date of the ride  
* season -  1 = spring, 2 = summer, 3 = fall, 4 = winter 
* holiday - whether the day is considered a holiday
* workingday - whether the day is neither a weekend nor holiday
* weather - 1: Clear, Few clouds, Partly cloudy, Partly cloudy
            2: Mist + Cloudy, Mist + Broken clouds, Mist + Few clouds, Mist
            3: Light Snow, Light Rain + Thunderstorm + Scattered clouds, Light Rain + Scattered clouds
            4: Heavy Rain + Ice Pallets + Thunderstorm + Mist, Snow + Fog 
* temp - temperature in Celsius
* atemp - "feels like" temperature in Celsius
* humidity - relative humidity
* windspeed - wind speed
* casual - number of non-registered user rentals initiated
* registered - number of registered user rentals initiated
* count - number of total rentals

In [8]:
import pandas as pd
import numpy as np 


In [9]:
pd.show_versions()



INSTALLED VERSIONS
------------------
commit           : e8093ba372f9adfe79439d90fe74b0b5b6dea9d6
python           : 3.9.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.3.0
Version          : Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.4.3
numpy            : 1.23.2
pytz             : 2022.2.1
dateutil         : 2.8.2
setuptools       : 62.6.0
pip              : 22.3.1
Cython           : 0.29.32
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.5.0
pandas_datareader: None
bs4              : Non

In [10]:
dataset_1 = pd.read_csv("/Users/michaeldionne/Library/CloudStorage/Dropbox/AI_ML Bootcamp/Caltech-AI-Machine-Learning-Bootcamp/Course4_Applied Data Science with Python/Datasets_4754/LVC Datasets/Lesson_5/rental_bike_descr.csv")


In [11]:
type(dataset_1)


pandas.core.frame.DataFrame

In [12]:
dataset_1.shape



(610, 10)

In [13]:
dataset_1.head()


instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   
2        3  01-01-2011       1   0     1   2    False        6           1   
3        4  01-01-2011       1   0     1   3    False        6           1   
4        5  01-01-2011       1   0     1   4    False        6           1   

   temp  
0  0.24  
1  0.22  
2  0.22  
3  0.24  
4  0.24

In [17]:
dataset_2 = pd.read_csv("/Users/michaeldionne/Library/CloudStorage/Dropbox/AI_ML Bootcamp/Caltech-AI-Machine-Learning-Bootcamp/Course4_Applied Data Science with Python/Datasets_4754/LVC Datasets/Lesson_5/rental_bike_season.csv")


In [18]:
dataset_2.shape

(610, 8)

In [19]:
dataset_2.head()


Unnamed: 0  instant   atemp   hum  windspeed  casual  registered  cnt
0           0        1  0.2879  0.81        0.0       3          13   16
1           1        2  0.2727  0.80        0.0       8          32   40
2           2        3  0.2727  0.80        0.0       5          27   32
3           3        4  0.2879  0.75        0.0       3          10   13
4           4        5  0.2879  0.75        0.0       0           1    1

In [21]:
dataset_2 = dataset_2.drop(['Unnamed: 0'], axis =1)


In [22]:
dataset_2.shape


(610, 7)

In [23]:
dataset_2.head()


instant   atemp   hum  windspeed  casual  registered  cnt
0        1  0.2879  0.81        0.0       3          13   16
1        2  0.2727  0.80        0.0       8          32   40
2        3  0.2727  0.80        0.0       5          27   32
3        4  0.2879  0.75        0.0       3          10   13
4        5  0.2879  0.75        0.0       0           1    1

In [24]:
combined_data = pd.merge(dataset_1, dataset_2, on="instant")


In [25]:
combined_data.shape


(610, 16)

In [26]:
combined_data.head()


instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   
2        3  01-01-2011       1   0     1   2    False        6           1   
3        4  01-01-2011       1   0     1   3    False        6           1   
4        5  01-01-2011       1   0     1   4    False        6           1   

   temp   atemp   hum  windspeed  casual  registered  cnt  
0  0.24  0.2879  0.81        0.0       3          13   16  
1  0.22  0.2727  0.80        0.0       8          32   40  
2  0.22  0.2727  0.80        0.0       5          27   32  
3  0.24  0.2879  0.75        0.0       3          10   13  
4  0.24  0.2879  0.75        0.0       0           1    1

In [27]:
dataset_3 = pd.read_csv("/Users/michaeldionne/Library/CloudStorage/Dropbox/AI_ML Bootcamp/Caltech-AI-Machine-Learning-Bootcamp/Course4_Applied Data Science with Python/Datasets_4754/LVC Datasets/Lesson_5/final_rental_bike_dataset.csv")


In [28]:
dataset_3.shape


(390, 16)

In [29]:
dataset_3.head()



instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0      620  29-01-2011       1   0     1   1    False        6           1   
1      621  29-01-2011       1   0     1   2    False        6           1   
2      622  29-01-2011       1   0     1   3    False        6           1   
3      623  29-01-2011       1   0     1   4    False        6           1   
4      624  29-01-2011       1   0     1   6    False        6           1   

   temp   atemp   hum  windspeed  casual  registered  cnt  
0  0.22  0.2273  0.64     0.1940       0          20   20  
1  0.22  0.2273  0.64     0.1642       0          15   15  
2  0.20  0.2121  0.64     0.1343       3           5    8  
3  0.16  0.1818  0.69     0.1045       1           2    3  
4  0.16  0.1818  0.64     0.1343       0           2    2

In [30]:
dataset_3.tail(15)


instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
375      995  14-02-2011       1   0     2   2    False        1           1   
376      996  14-02-2011       1   0     2   3    False        1           1   
377      997  14-02-2011       1   0     2   4    False        1           1   
378      998  14-02-2011       1   0     2   5    False        1           1   
379      999  14-02-2011       1   0     2   6    False        1           1   
380     1000  14-02-2011       1   0     2   7    False        1           1   
381      611  28-01-2011       1   0     1  16    False        5           1   
382      612  28-01-2011       1   0     1  17    False        5           1   
383      613  28-01-2011       1   0     1  18    False        5           1   
384      614  28-01-2011       1   0     1  19    False        5           2   
385      615  28-01-2011       1   0     1  20    False        5           2   
386      616  28-01-2011       1   0     1  21    False        5           2   
387      617  28-01-2011       1   0     1  22    False        5           1   
388      618  28-01-2011       1   0     1  23    False        5           1   
389      619  29-01-2011       1   0     1   0    False        6           1   

     temp   atemp   hum  windspeed  casual  registered  cnt  
375  0.36  0.3333  0.40     0.2985       0           2    2  
376  0.34  0.3182  0.46     0.2239       1           1    2  
377  0.32  0.3030  0.53     0.2836       0           2    2  
378  0.32  0.3030  0.53     0.2836       0           3    3  
379  0.34  0.3030  0.46     0.2985       1          25   26  
380  0.34  0.3030  0.46     0.2985       2          96   98  
381  0.22  0.2727  0.80     0.0000      10          70   80  
382  0.24  0.2424  0.75     0.1343       2         147  149  
383  0.24  0.2273  0.75     0.1940       2         107  109  
384  0.24  0.2424  0.75     0.1343       5          84   89  
385  0.24  0.2273  0.70     0.1940       1          61   62  
386  0.22  0.2273  0.75     0.1343       1          57   58  
387  0.24  0.2121  0.65     0.3582       0          26   26  
388  0.24  0.2273  0.60     0.2239       1          22   23  
389  0.22  0.1970  0.64     0.3582       2          26   28

In [31]:
#sort values by columns
dataset_3 = dataset_3.sort_values(by="instant")


In [32]:
dataset_3.head()


instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
381      611  28-01-2011       1   0     1  16    False        5           1   
382      612  28-01-2011       1   0     1  17    False        5           1   
383      613  28-01-2011       1   0     1  18    False        5           1   
384      614  28-01-2011       1   0     1  19    False        5           2   
385      615  28-01-2011       1   0     1  20    False        5           2   

     temp   atemp   hum  windspeed  casual  registered  cnt  
381  0.22  0.2727  0.80     0.0000      10          70   80  
382  0.24  0.2424  0.75     0.1343       2         147  149  
383  0.24  0.2273  0.75     0.1940       2         107  109  
384  0.24  0.2424  0.75     0.1343       5          84   89  
385  0.24  0.2273  0.70     0.1940       1          61   62

In [33]:
dataset_3.tail()


instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
376      996  14-02-2011       1   0     2   3    False        1           1   
377      997  14-02-2011       1   0     2   4    False        1           1   
378      998  14-02-2011       1   0     2   5    False        1           1   
379      999  14-02-2011       1   0     2   6    False        1           1   
380     1000  14-02-2011       1   0     2   7    False        1           1   

     temp   atemp   hum  windspeed  casual  registered  cnt  
376  0.34  0.3182  0.46     0.2239       1           1    2  
377  0.32  0.3030  0.53     0.2836       0           2    2  
378  0.32  0.3030  0.53     0.2836       0           3    3  
379  0.34  0.3030  0.46     0.2985       1          25   26  
380  0.34  0.3030  0.46     0.2985       2          96   98

In [34]:
#concatenate
final_data = pd.concat([dataset_3, combined_data])


In [35]:
final_data.shape


(1000, 16)

In [37]:
final_data = final_data.rename(columns ={'dteday': 'date', 'yr': 'year', 'mnth':'month','hr':'hour', 'weathersit':'weather', 'hum':'humidity', 'cnt':'count'})



In [38]:
final_data.head()


instant        date  season  year  month  hour  holiday  weekday  \
381      611  28-01-2011       1     0      1    16    False        5   
382      612  28-01-2011       1     0      1    17    False        5   
383      613  28-01-2011       1     0      1    18    False        5   
384      614  28-01-2011       1     0      1    19    False        5   
385      615  28-01-2011       1     0      1    20    False        5   

     weather  temp   atemp  humidity  windspeed  casual  registered  count  
381        1  0.22  0.2727      0.80     0.0000      10          70     80  
382        1  0.24  0.2424      0.75     0.1343       2         147    149  
383        1  0.24  0.2273      0.75     0.1940       2         107    109  
384        2  0.24  0.2424      0.75     0.1343       5          84     89  
385        2  0.24  0.2273      0.70     0.1940       1          61     62

In [39]:
final_data.dtypes


instant         int64
date           object
season          int64
year            int64
month           int64
hour            int64
holiday          bool
weekday         int64
weather         int64
temp          float64
atemp         float64
humidity      float64
windspeed     float64
casual          int64
registered      int64
count           int64
dtype: object

In [40]:
final_data.isna()


instant   date  season   year  month   hour  holiday  weekday  weather  \
381    False  False   False  False  False  False    False    False    False   
382    False  False   False  False  False  False    False    False    False   
383    False  False   False  False  False  False    False    False    False   
384    False  False   False  False  False  False    False    False    False   
385    False  False   False  False  False  False    False    False    False   
..       ...    ...     ...    ...    ...    ...      ...      ...      ...   
605    False  False   False  False  False  False    False    False    False   
606    False  False   False  False  False  False    False    False    False   
607    False  False   False  False  False  False    False    False    False   
608    False  False   False  False  False  False    False    False    False   
609    False  False   False  False  False  False    False    False    False   

      temp  atemp  humidity  windspeed  casual  registered  count  
381  False  False     False      False   False       False  False  
382  False  False     False      False   False       False  False  
383  False  False     False      False   False       False  False  
384  False  False     False      False   False       False  False  
385  False  False     False      False   False       False  False  
..     ...    ...       ...        ...     ...         ...    ...  
605  False  False     False      False   False       False  False  
606  False  False     False      False   False       False  False  
607  False  False     False      False   False       False  False  
608  False  False     False      False   False       False  False  
609  False  False     False      False   False       False  False  

[1000 rows x 16 columns]

The isna().sum(axis=0) function provides a clear picture of the number of null values in a DataFrame
In the given result, we can see that the atemp column has 11 null values

In [41]:
final_data.isna().sum(axis=0)


instant        0
date           0
season         0
year           0
month          0
hour           0
holiday        0
weekday        0
weather        0
temp           0
atemp         11
humidity       0
windspeed      0
casual         0
registered     0
count          0
dtype: int64

In [42]:
percentage_of_missing_values = (final_data['atemp'].isna().sum(axis=0)/final_data.shape[0])*100
percentage_of_missing_values


1.0999999999999999

We divide the number of null values by the shape of the DataFrame to get the percentage of missing values.
Since the percentage is 1, it is very less. Usually, the industry practice allows us to drop rows up to 30%. So, we can drop the rows with
missing values.

In [43]:
final_data = final_data.dropna(axis=0)
final_data.shape


(989, 16)

We can see that the shape of the DataFrame reduced to 989 from 1000. It shows that the missing value rows have been wiped off.

In [44]:
final_data.isna().sum(axis=0)


instant       0
date          0
season        0
year          0
month         0
hour          0
holiday       0
weekday       0
weather       0
temp          0
atemp         0
humidity      0
windspeed     0
casual        0
registered    0
count         0
dtype: int64

In [45]:
np.sum(final_data['casual'] + final_data['registered'] - final_data['count'])


0

In [47]:
# to find the elements of an array
np.unique(final_data.month)


array([1, 2])

In [48]:
np.unique(final_data.hour)


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [49]:
print(final_data.describe())


           instant  season   year       month        hour     weekday  \
count   989.000000   989.0  989.0  989.000000  989.000000  989.000000   
mean    505.622851     1.0    0.0    1.315470   11.753286    2.991911   
std     286.274765     0.0    0.0    0.464938    6.891129    2.084727   
min       1.000000     1.0    0.0    1.000000    0.000000    0.000000   
25%     259.000000     1.0    0.0    1.000000    6.000000    1.000000   
50%     506.000000     1.0    0.0    1.000000   12.000000    3.000000   
75%     753.000000     1.0    0.0    2.000000   18.000000    5.000000   
max    1000.000000     1.0    0.0    2.000000   23.000000    6.000000   

          weather        temp       atemp    humidity   windspeed      casual  \
count  989.000000  989.000000  989.000000  989.000000  989.000000  989.000000   
mean     1.479272    0.204712    0.211958    0.581769    0.194609    4.921132   
std      0.651085    0.077789    0.076703    0.187706    0.129225    7.666231   
min      1.000000 